### PR TITLE
Integration with the upgrade service

### PR DESCRIPTION
**Pull request by dianabarsan**
# Description

Adds API calls to the upgrade service. 
Updates admin app to reload the upgrade page when upgrade is successful. 
Publishes docker-compose file and tags on the staging server. 
Adds support for npm 8.6

medic/cht-core#7571

# Code review checklist

- [ ] Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- [ ] Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- [ ] Tested: Unit and/or e2e where appropriate
- [ ] Internationalised: All user facing text
- [ ] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes.

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -86,7 +86,7 @@ jobs:
         node-version: 16.x
     - name: Grunt Install
       run: npm install -g grunt-cli
-    - run: npm ci
+    - run: npm ci --legacy-peer-deps
```
Comment (dianabarsan) [inline .github/workflows/build.yml:89]: This is temporary until we property resolve dependencies for 4.0

**Comment by dianabarsan** [comment]
Please have a look, @garethbowen . Thanks a lot!

File: admin/src/js/controllers/upgrade.js [hunk 2]
```diff
@@ -164,13 +178,30 @@ angular.module('controllers').controller('UpgradeCtrl',
       }).catch(() => {});
     };
 
+    const waitUntilApiStarts = () => {
+      return $http
+        .get(POLL_URL)
+        .catch(() => {
+          const waitOneSecond = () => new Promise(r => $timeout(r, 1000));
+          return waitOneSecond().then(() => waitUntilApiStarts());
```
Comment (garethbowen) [inline admin/src/js/controllers/upgrade.js:186]: Will this make an increasingly long promise chain? Can we use a callback, (ie: pass in a Promise.resolve function) so we don't keep chaining?

File: api/src/services/setup/utils.js [hunk 3]
```diff
@@ -291,6 +294,42 @@ const unstageStagedDdocs = async () => {
...
+const makeUpgradeRequest = (payload) => {
+  let url;
+  try {
+    url = new URL(UPGRADE_SERVICE_URL);
+    url.pathname = '/upgrade';
+  } catch (err) {
+    throw new Error(`Invalid UPGRADE_SERVICE_URL: ${UPGRADE_SERVICE_URL}`);
+  }
+
+  return rpn.post({ url: `${UPGRADE_SERVICE_URL}/upgrade`, json: true, body: payload });
```
Comment (garethbowen) [inline api/src/services/setup/utils.js:330]: Surely you can use the URL you created above?

```suggestion
  return rpn.post({ url, json: true, body: payload });
```

**Review by garethbowen — CHANGES_REQUESTED** [review]
Awesome work.

A few minor things inline, and also...

1. It's getting complicated with the sheer number of docker-compose files floating around, and this PR adds a templated one. How will we keep them all in sync? Can other ones be removed and replaced with one generated off this template?
2. I presume all the package-lock changes were due to bumping npm version?

File: api/src/services/setup/utils.js (hunk 3 above)
Comment (dianabarsan) [reply api/src/services/setup/utils.js:330]: Oh, that was the intention .. and then life happened :)) 
Thanks for noticing!

File: .github/workflows/build.yml [hunk 4]
```diff
@@ -101,6 +101,18 @@ jobs:
       with:
         name: cht-images
       if: ${{ !env.INTERNAL_CONTRIBUTOR }}
+    - name: Configure AWS credentials
+      uses: aws-actions/configure-aws-credentials@v1
+      with:
+        aws-access-key-id: ${{ secrets.ECR_ACCESS_KEY_ID }}
```
[the remark this answers is not in the document]
Comment (dianabarsan) [reply .github/workflows/build.yml:107]: I don't think this has to be a secret. Do you think we should hardcode this in the code somewhere? Historically we had trouble with that (chrome version with protractor). 
I'm not sure if there's a way to store a non-secret workflow variable outside of the repository code (https://github.community/t/unencrypted-environment-variables-like-not-secrets/186988/2)

**Comment by dianabarsan** [comment]
> It's getting complicated with the sheer number of docker-compose files floating around, and this PR adds a templated one. How will we keep them all in sync? Can other ones be removed and replaced with one generated off this template?

I've changed it so I create the test docker-compose at runtime. It complicated the template a little. 

> I presume all the package-lock changes were due to bumping npm version?

Yea, the version fixed in the package.json didn't match the version in package-lock. 
As of yesterday afternoon, the version of npm installed in CI was 8.5, and it fails the install with these inconsistencies. 
I could try to for an install of npm 8.3 and leave the package-lock changes for another PR.

**Comment by dianabarsan** [comment]
> It's getting complicated with the sheer number of docker-compose files floating around, and this PR adds a templated one. How will we keep them all in sync? Can other ones be removed and replaced with one generated off this template?

I've changed it so I create the test docker-compose at runtime. It complicated the template a little. 

> I presume all the package-lock changes were due to bumping npm version?

Yea, the version fixed in the package.json didn't match the version in package-lock. 
As of yesterday afternoon, the version of npm installed in CI was 8.5, and it fails the install with these inconsistencies. 
I could try to for an install of npm 8.3 and leave the package-lock changes for another PR.

**Review by garethbowen — APPROVED** [review]
Loving it!

Approved pending one optional change about whether the AWS user is secret or hardcoded.

**Pull request merged**